python/llm/example/GPU/Deepspeed-AutoTP-FastAPI/serving.py

```diff
@@ -89,70 +115,144 @@ def load_model(model_path, low_bit):
+        skip_prompt=True,
+        skip_special_tokens=True,
+        batch_size=len(prompt),
+    )
 
+    generated_kwargs = dict(
+        max_new_tokens=n_predict,
+        streamer=streamer,
+        attention_mask=attention_mask,
+    )
 
-class PromptRequest(BaseModel):
-    prompt: str
-    n_predict: int = 32  
+    def model_generate():
+        model.generate(input_ids, **generated_kwargs)
+        torch.xpu.synchronize()
+
+    t1 = Thread(target=model_generate)
+    t1.start()
+
+    partial_output = ""
+    rfind_start = 0
+
+    async def put_item(queue, item):
+        await queue.put(item)
+
+    for i in range(n_predict):
+        tasks = []
+        output_token = next(streamer)
+        for index, request_id in enumerate(request_ids):
+            task = asyncio.create_task(
+                put_item(
+                    streamer_dict[request_id], (n_predict - 1 - i, output_token[index])
+                )
+            )
+            tasks.append(task)
+        await asyncio.gather(*tasks)
 
-empty_req = PromptRequest(prompt="", n_predict=0)
 
 app = FastAPI()
 
-from collections import deque
-rest_req_deque = deque(maxlen=128)
-request_queue: asyncio.Queue = asyncio.Queue()
-result_dict: Dict[str, str] = {}
+
+async def stream_generator(token_queue, request_id):
+    while True:
+        if not token_queue.empty():
+            remain, token = await token_queue.get()
+            yield json.dumps({"generate_text": token}) + "\n"
+            if remain == 0:
+                break
+        else:
+            await asyncio.sleep(0.001)
+    streamer_dict.pop(request_id, None)
+
+
+async def generator(token_queue, request_id):
+    while True:
+        if not token_queue.empty():
+            remain, token = await token_queue.get()
+            yield token
+            if remain == 0:
+                break
+        else:
+            await asyncio.sleep(0.01)
+    streamer_dict.pop(request_id, None)
+
 
 @app.post("/generate/")
 async def generate(prompt_request: PromptRequest):
     request_id = str(uuid.uuid4())
     await request_queue.put((request_id, prompt_request))
     while True:
         await asyncio.sleep(0.1)
-        if request_id in result_dict:
-            output_str = result_dict.pop(request_id)
-            return {"generated_text": output_str}
+        if request_id in streamer_dict and not request_id.endswith("stream"):
+            output_str = []
+            token_queue = streamer_dict[request_id]
+            async for item in generator(token_queue, request_id):
+                output_str.append(item)
+
+            return {"generated_text": "".join(output_str)}
+
+
+@app.post("/generate_stream/")
+async def generate_stream(prompt_request: PromptRequest):
+    request_id = str(uuid.uuid4()) + "stream"
+    await request_queue.put((request_id, prompt_request))
+    while True:
+        await asyncio.sleep(0.1)
+        if request_id in streamer_dict and request_id.endswith("stream"):
```

Comment:
Is the check ` request_id.endswith("stream")` necessary? `request_id` should be init in line 245 and not change in the function?